### ISSUE TITLE
[Feature Request] Remove Analog Stick Ingame Deadzone

### ISSUE THREAD

**Issue by jubei867** [hifihedgehog/PadForge#50]
### Description

Some games seem to have a deadzone around the center of an analog stick. This can be annoying when using various input devices. So it would be helpful if PadForge had an "inverted deadzone" adjuster for the input device.

Smth like: when setting the inverted deadzone to 5%, the virtual controller always skips the first 5% of input aroud the center. So pressing 1% to the left of my input device will result in pressing 6% to the left of the virtual controller.

### Use Case

I play arcade racing like "circuit superstars" and "art of rally". I play with steering wheel and pedals as a virtual controller. When doing so, there is a noticeable deadzone around the steering center. This ruins the feeling quite a bit. 

Having a feature like I described, I would try to set the inverted deadzone to a value, where this deadzone gets so little that it is hardly noticebale anymore.

**Comment by hifihedgehog** [comment]
Hi @jubei867, great to have you here, and thanks for the suggestion! Good news — PadForge already has exactly this feature built in. It's called **Anti-Dead Zone**, and you'll find it in your virtual controller's **Sticks tab** (open a controller slot, then go to Sticks). There's a per-axis Anti-Dead Zone X / Y slider sitting right next to the regular Dead Zone X / Y.

The wiki has a full write-up on how it works here: **[Stick Dead Zones → Anti-Dead Zone](https://github.com/hifihedgehog/PadForge/wiki/Stick-Dead-Zones#anti-dead-zone)**. The short version, straight from the wiki:

> Many games have built-in dead zones you cannot disable. Anti-Dead Zone compensates by guaranteeing a minimum output the instant the stick crosses PadForge's dead zone boundary.
>
> **Formula:** `output = antiDeadZone + (stickInput * (1 - antiDeadZone))`

**One important catch:** Anti-Dead Zone needs a **non-zero regular Dead Zone** to anchor to. Without a small reference dead zone, the anti-dead zone has no boundary to "jump from," and you'll get the stick snapping to unexpected positions instead of the smooth offset you want. The wiki explains it like this:

> | Stick Position | Output |
> |---|---|
> | Inside dead zone | 0% |
> | Just past dead zone | Jumps to 20% |
> | Full deflection | 100% |
> | In between | Spread across 20–100% |

So you'll want to play with **both values together**, since your physical wheel will also have its own native dead zone you can use in combination. A practical starting point for racing wheels (also from the wiki):

> ### Racing Games
> | Setting | Steering |
> |---|---|
> | Dead Zone | 3–8% |
> | Anti-Dead Zone | start small, increase 5% at a time |

The wiki has a recommended workflow:

> 1. Set the dead zone just high enough to eliminate drift.
> 2. Slowly push the stick and watch the preview dot. If the dot moves but nothing happens in-game, the game's dead zone is eating your input.
> 3. Increase anti-dead zone in 5% increments until the game responds immediately.
> 4. Stop once the game responds — going higher compresses your usable range.

There's also a **Linear** slider on the same page that blends between the anti-dead-zone curve and a fully linear response, which can help fine-tune the feel once you've got the two dead zones dialed in.

Give it a try and let me know how it works for Circuit Superstars and Art of Rally — happy to help you dial it in if anything feels off!

**Comment by jubei867** [comment]
That's amazing, now im rocking these games ;D
... well, I don't ^^ But the car handling is on point now and makes a huge difference!

But it doesn't seem to work properly. I can set an anti-deadzone. I can see the anti-deadzone in the gamepad preview - the analog stick skips this part. And it's working ingame. But as soon as I change a setting in PadForge, the physical appearance of the anti-deadzone vanishes. That means no anti-deadzone ingame and in the gamepad preview in PadForge. Changing settings does not bring it back. But a PadForge restart at least does.

**Comment by hifihedgehog** [comment]
Glad to hear the car handling is on point now — that's exactly the kind of difference anti-deadzone makes!

I tried reproducing the bug you described (anti-deadzone vanishing after changing a setting) but wasn't able to on my end — it persists correctly through settings changes for me. Could you let me know what version of PadForge you're running? That'll help me narrow down what's going on. You can find the version number in Settings at the bottom of the page.

**Comment by jubei867** [comment]
Thanks for looking after my issue! 
I guess a video of the issue tells more than my english ^^
https://youtu.be/2KmAaJjOYMQ

Note: At the end of the video, after I restart PadForge, I don't change a setting at all, just tabbed to the Stick settings and back, then the Anti-Deadzone vanished already.

I'm on v2.3.0

**Comment by hifihedgehog** [comment]
Hey @jubei867. I noticed that you said you are 2.3.0. However, after looking at the video, it is clear you used interday builds (the ones with the random numbers like r682@12f82ba). Unlike releases, these are not necessarily as guaranteed to work, they occur *anytime* I do a commit including builds throughout a day of development including sometimes when I am in the middle of making changes, they are **_never_** guaranteed to work, and you could have very well picked up a buggy file in this way. I would need to see the actual archive version to try to do a trace on this and determine if this is still a problem as I may be wiring something in and you picked up something that wasn't right. As it is, if I switch tabs and decrease the anti-dead zone as you did and am not seeing it with the latest non-release build so I am guessing you picked up something while I was in the middle of development, maybe? Have you tried that by chance?

**Comment by jubei867** [comment]
That's true, I picked the wrong version for the video. Stupid me, sorry. But sadly, it doesn't change a thing. It also happens with version 2.3.0 and v2.4.0, too.

Some more weird behaviour I observed:
I have 2 usb devices added to the virtual controller: the wheelbase and the pedals. When the issue occurs, i.e. the physical appearance of the anti-deadzone vanished, I can see, that PadForge has zeroed the anti-deadzone value, by going to the Stick Tab and switch to the pedals device and back to the wheelbase device. There the setting for anti-deadzone is automatically set back to 0. (settings for normal deadzone stays fyi.)

Following this observation, I tried creating a virtual controller using only the wheelbase as an input device. But in this case I was not able to get the anti-deadzone to work a single time! I can set the value to what I want, there is no physical appearance of the anti-deadzone.

*** little bit off topic, but maybe not:
I'm also having another big issue with PadForge. In case these issues have smth to do with each other, I mention it quickly: When PadForge ist running, even with no virtual controller active, my keyboard inputs are getting messed up. Sometimes an input will be skipped, sometimes the last input will be repeated until a new input is made.
This happens to me in all three PadForge versions I had obviously used.
I wanted to narrow this issue a little bit more down, bevor opening a new issue on here. Hope to be able to tell more about this soon...

**Comment by jubei867** [comment]
I made some more tests and have new results:

I just tried the same scenario an another system, a win10 laptop. Installed the newest PadForge Version. Having the exact same issue!

Then I also tried creating a virtual xbox controller controlled by a logitech usb controller instead of the fanatec wheel. I use an analog stick of the logitech controller to mimic the xbox stick. Then I set up an anti-deadzone. And the same issue occurs.

So I reproduced the error using completely different hardware!

The best way, to have the anti-deadzone working for me, is to go to the Stick-Tab in PadForge, set the desired value for anti-deadzone, and then restart PadForge without clicking anything else in PadForge before restarting. 
This seems to assure the anti-deadzone.
But as soon as I reenter the stick tab, the error has a good chance to occur. I.e. the anti-deadzone can be vanished already at this point, even if the setting shows something different. 

(the offtopic thing about keyboard input issues can be ignored by now. On my laptop I didn't see this issue happening, so it seems to have nothing to do with the anti-deadzone issue.)

**Comment by hifihedgehog** [comment]
> The best way, to have the anti-deadzone working for me, is to go to the Stick-Tab in PadForge, set the desired value for anti-deadzone, and then restart PadForge without clicking anything else in PadForge before restarting.
> This seems to assure the anti-deadzone.

I tried that and no dice. However, it dawned on me. Do you have a shifter or similar set to the left stick Y axis under a different device that makes up your virtual controller? Something that moves the joystick up and down under a different device? If so, *THAT'S" it. You have to use the axial setting. The reason being is scaled radial deadzone prevents any deadzone or anti-deadzone above or below the y axis. Scaled radial is 9 times out of 10 the right one to use because you don't want the stick locking to the orthogonal directions (that is up, down, left, and right) when you are walking or moving around in a shooter or action-adventure title. However, with a driving game, you want that axial deadzone in order to always be hopping across you x-axis deadzone of your steering no matter how far away you are from center up or down especially with a steering wheel that has a shifter on Y. Is that it?

@jubei867

**Comment by hifihedgehog** [comment]
https://github.com/user-attachments/assets/0d68486e-062c-4869-88c7-e14822b33020

As it is, though, under my conditions, as you can see, I simply can't replicate this.

**Comment by hifihedgehog** [comment]
@jubei867, I also tried having two devices connected with similar deadzone and anti-deadzone values set and tried tracing your video almost exactly, choreographically speaking, especially [from 1:22 to end](https://youtu.be/2KmAaJjOYMQ?si=zPVlLWryRspBQ6jG&t=82). The topmost device I saw in your virtual controller has no deadzone or anti-deadzone values set (a slightly non-zero Y deadzone and a mild zero X anti-deadzone) , whereas the video has some non-zero values set in deadzone and anti deadzone.

<img width="565" height="396" alt="Image" src="https://github.com/user-attachments/assets/05e53706-eaac-4eb4-b63b-0c7833309cb1" />

<img width="552" height="405" alt="Image" src="https://github.com/user-attachments/assets/92201751-191f-40dc-a78b-2cc3d03498b0" />

<img width="570" height="403" alt="Image" src="https://github.com/user-attachments/assets/a9eaa2cd-ed20-4255-9b2b-40374ae935eb" />

<img width="577" height="411" alt="Image" src="https://github.com/user-attachments/assets/3ccb4a6b-0884-429a-9bff-de7d9052fdd7" />

<img width="572" height="411" alt="Image" src="https://github.com/user-attachments/assets/941f86b5-8e90-4aaf-8555-628d19bbf2d6" />

<img width="568" height="392" alt="Image" src="https://github.com/user-attachments/assets/8f916779-c63d-4e25-a7a2-041092392c0d" />

I tried doing that following that in your video, and then after closing, I followed the mouse actions and even the X axis position of your wheel (light-medium positive X or right leaning left stick axis at launch) to approximate what you are seeing. The GUI path you followed in the second launch was click on the first virtual controller's card in the dashboard page, click on the sticks tab which lands on a device with "0" on the deadzone and anti-deadzone values, click on the assigned devices dropdown and select the wheel which still shows its seettings, and then click on the preview tab. I then start to roll my X axis from right to left slowly but I inevitably see it hop over center as expected and not smoothly through center as your video shows.

I cannot replicate it no matter how hard I try. As a last-ditch effort, could you send me your .xml file? That should be more telling of what you have going on here. There could be something in your settings that is throwing things off or represents a unique corner case that I have not yet encountered.

**Comment by hifihedgehog** [comment]
@jubei867 — great news, I think we've found it, and it's not something you'd ever guess from the symptoms.

This is a **locale / decimal separator bug**. Your Windows is set to German (I noticed PadForge's UI is showing "Übersicht" in your screenshots, which tipped me off). German Windows uses a **comma** as the decimal separator instead of a dot — so where English Windows writes `20.5`, German Windows writes `20,5`.

PadForge has an asymmetry in how it writes vs. reads these values:

- **Saving** uses culture-sensitive `.ToString()`, which on your German system writes anti-deadzone values as `"20,5"` (with a comma).
- **Loading** correctly uses `InvariantCulture` (the right thing for serialized data), which expects `"20.5"` (with a dot).

The result: every time PadForge saves your anti-deadzone, it writes it with a comma. Then the next time anything triggers a reload of the values into the UI — switching tabs, switching devices, re-opening the Sticks page — the parser sees `"20,5"`, fails to recognize it, and silently falls back to `0`. There's a 30Hz background save loop that then writes that `0` straight back to the file, making the corruption permanent.

This explains every detail of what you observed:

- ✅ **Reproduced on two completely different systems with different hardware** — same German locale on both
- ✅ **I couldn't reproduce on en-US** — `20.5` writes and reads as `20.5`, no separator confusion
- ✅ **Deadzone survives but anti-deadzone doesn't** — if your deadzone values happen to be whole integers, they round-trip cleanly in any locale; anti-deadzone tends to land on fractional values from slider dragging
- ✅ **Setting it, then restarting PadForge, makes it stick once** — until the next save cycle re-corrupts it
- ✅ **Tabbing into Sticks and back zeros it** — that triggers a reload, which trips the parser
- ✅ **The slider visually shows your value but the runtime uses 0** — the UI control briefly holds the right number until the next reload pass clobbers it

This is also almost certainly affecting **other settings** beyond just anti-deadzone — anything fractional that gets saved through the same path. You may have been seeing weird behavior in other places too without realizing it had the same root cause.

Thank you for sticking with me on this — your willingness to test on a second machine with completely different hardware was the key clue that ruled out everything else and pointed me toward something environmental rather than hardware. The fix is straightforward (force `InvariantCulture` on the save side everywhere a number gets serialized) and I'll get it into the next build. I'll ping you here when there's something to test.

**Comment by hifihedgehog** [comment]
One quick follow-up — since I'm in the middle of a major v3.0.0 migration and don't want to fragment fixes across branches, the proper fix is going to land in v3.0.0 rather than as a v2.4.2 hotfix. In the meantime, here are two workarounds you can use right now to get anti-deadzone working on v2.4.1:

### Option 1 — Temporarily switch Windows decimal separator (easiest)

This is the cleanest workaround and takes about 30 seconds:

1. Open Windows **Settings → Time & Language → Region**
2. Click **Additional date, time & regional settings** → **Region** → **Additional settings...** (or **Format settings**)
3. On the **Numbers** tab, change **Decimal symbol** from `,` to `.`
4. Click **OK**, then **restart PadForge**

You can leave the rest of your German formatting alone — only the decimal separator needs to change. After this, PadForge will save and load anti-deadzone values correctly, and you can use the Sticks tab freely without anything getting zeroed out. When v3.0.0 ships with the proper fix, you can switch your decimal separator back to `,` and everything will keep working.

### Option 2 — Manually edit PadForge.xml (more fragile)

If you don't want to touch your Windows settings, you can edit `PadForge.xml` directly:

1. Close PadForge
2. Open `PadForge.xml` in a text editor (it's in the same folder as `PadForge.exe`)
3. Find the `<LeftThumbAntiDeadZoneX>`, `<LeftThumbAntiDeadZoneY>`, `<RightThumbAntiDeadZoneX>`, `<RightThumbAntiDeadZoneY>` (and trigger anti-deadzone) entries for your wheel device
4. Replace any `,` with `.` in those values (e.g., `20,5` → `20.5`)
5. Save the file and start PadForge
6. **Don't open the Sticks tab** — any tab navigation will trigger a reload that re-corrupts the values

Option 1 is much less fragile and I'd strongly recommend it.

I'll ping you here when v3.0.0 has a testable build with the proper fix.

**Comment by jubei867** [comment]
That's it! Problem solved!

As you said, when using only integers, the anti-deadzone is working and doesn't vanish.
And switching the Windows decimal seperator to ".", the anti-deadzone is working for me with every decimal number.

That's amazing! Thank you!

**Comment by hifihedgehog** [comment]
Fixed in v3-dev for the forthcoming v3.0.0 release:

- [`928d5ff`](https://github.com/hifihedgehog/PadForge/commit/928d5ff24a0af3570d4883cd0ac94ea73b328e9d) — use `InvariantCulture` for all double→string serialization, so dead zone / anti-dead zone values round-trip correctly regardless of your Windows locale
- [`edfd0a0`](https://github.com/hifihedgehog/PadForge/commit/edfd0a06c1a1465c89468f1564a9b0ddaa0a6cef) — developer warning comments to prevent this class of bug from regressing

No v2.x backport — this ships in v3.0.0. Thanks again for the careful repro and the second-system test that pinned it to locale.

**Comment by hifihedgehog** [comment]
This issue is closed and locked. If you hit what looks like the same problem on a current release, please file a **fresh issue** rather than commenting here — comments on locked threads do not surface in triage, and the original diagnosis may not apply to a recurrence even if the symptoms look similar.

Quick checklist before filing:

- Confirm you are on the latest release.
- Reproduce in a clean folder (no leftover `PadForge.xml`).
- Capture `crash.log` if one exists next to `PadForge.exe`.
- Note your Windows build (`winver`) and your controller's exact model.
- Mention this issue number for context if it seems related, but do not assume the prior fix applies.

Thanks for using PadForge.
